src/librustc_resolve/macros.rs

```diff
@@ -558,12 +607,14 @@ impl<'a> Resolver<'a> {
                         this.graph_root.unexpanded_invocations.borrow().is_empty()
                     )),
                 }
-                Scope::ToolPrelude => if KNOWN_TOOLS.contains(&ident.name) {
-                    let binding = (Res::ToolMod, ty::Visibility::Public, DUMMY_SP, ExpnId::root())
-                                   .to_name_binding(this.arenas);
-                    Ok((binding, Flags::PRELUDE))
-                } else {
-                    Err(Determinacy::Determined)
+                Scope::ToolPrelude => match this.registered_tools.get(&ident).cloned() {
+                    Some(ident) => {
+                        let binding = (Res::ToolMod,
+                                       ty::Visibility::Public, ident.span, ExpnId::root())
+                                       .to_name_binding(this.arenas);
+                        Ok((binding, Flags::PRELUDE))
```

Comment:
Some of this logic is repeated above and in other places; could it be de-duplicated?